File: clients/Cpp/model/AttackAction.cpp
#include "AttackAction.hpp"

AttackAction::AttackAction() { }
AttackAction::AttackAction(std::shared_ptr<int> target, std::shared_ptr<AutoAttack> autoAttack) : target(target), autoAttack(autoAttack) { }
AttackAction AttackAction::readFrom(InputStream& stream) {
    AttackAction result;
    if (stream.readBool()) {
        result.target = std::shared_ptr<int>(new int());
        *result.target = stream.readInt();
    } else {
        result.target = std::shared_ptr<int>();
    }
    if (stream.readBool()) {
        result.autoAttack = std::shared_ptr<AutoAttack>(new AutoAttack());
        *result.autoAttack = AutoAttack::readFrom(stream);
    } else {
        result.autoAttack = std::shared_ptr<AutoAttack>();
    }
    return result;
}
void AttackAction::writeTo(OutputStream& stream) const {
    if (target) {
        stream.write(true);
        stream.write((*target));
    } else {
        stream.write(false);
    }
    if (autoAttack) {
        stream.write(true);
        (*autoAttack).writeTo(stream);
    } else {
        stream.write(false);
    }
}

File: clients/Scala/src/main/scala/model/AttackAction.scala
package model

import util.StreamUtil

case class AttackAction(target: Option[Int], autoAttack: Option[model.AutoAttack]) {
    def writeTo(stream: java.io.OutputStream) {
        target match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
        autoAttack match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                value.writeTo(stream)
            }
        }
    }
}
object AttackAction {
    def readFrom(stream: java.io.InputStream): AttackAction = AttackAction(
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None
        ,
        if (StreamUtil.readBoolean(stream)) Some(
            model.AutoAttack.readFrom(stream)
        ) else None
        )
}
